### ISSUE TITLE
API to register custom elements

### ISSUE THREAD

**Issue by jedrzejboczar** [rcarriga/nvim-dap-ui#154]
Hi, I wanted to extend dap-ui with a custom element for [nvim-dap-cortex-debug](https://github.com/jedrzejboczar/nvim-dap-cortex-debug) which can display RTT output in terminals. 

I was able to hook into nvim-dap-ui by creating a file under `lua/dapui/elements/rtt.lua`. This way dapui [can find the element](https://github.com/rcarriga/nvim-dap-ui/blob/1cd4764221c91686dcf4d6b62d7a7b2d112e0b13/lua/dapui/init.lua#L10), but the problem is that it won't [register](https://github.com/rcarriga/nvim-dap-ui/blob/1cd4764221c91686dcf4d6b62d7a7b2d112e0b13/lua/dapui/init.lua#L145) the element because it's not defined in [config.elements](https://github.com/rcarriga/nvim-dap-ui/blob/1cd4764221c91686dcf4d6b62d7a7b2d112e0b13/lua/dapui/config/init.lua#L3). If I add `RTT = "rtt"` then everything works fine. 

Is there any preferred mechanism for registering additional elements? I could directly write to config.elements but I would have to ensure that nvim-dap-cortex-debug setup is called before nvim-dap-ui setup which might be problematic. Would it be possible to defer elements registration until dapui is opened and add some API for registering elements? This way order of calls to setup() for nvim-dap-ui vs other plugins wouldn't matter.

Another related issue is that I have my element already in a buffer because it has to be created in response to an event from debug adapter, but dap-ui elements are expected to use the buffers that dap-ui provides. I've seen that the repl element has similar issue, and it is solved by [reassigning the current buffer to win_bufs](https://github.com/rcarriga/nvim-dap-ui/blob/c8ce83a66deb0ca6f5af5a9f9d5fcc05a6d0f66b/lua/dapui/windows/layout.lua#L43-L47). However this doesn't work when my element has been created in a different window and is not the current buffer. Currently I'm [replacing the buffer in layout.win_bufs](https://github.com/jedrzejboczar/nvim-dap-cortex-debug/blob/355e3d352ca42fd497876ae73300cc476cd9d226/lua/dapui/elements/rtt.lua#L32-L49). Maybe it would be better to allow `element.setup_buffer` to return a value `buf` - if it is not `nil` then this buffer would be used [here](https://github.com/rcarriga/nvim-dap-ui/blob/1cd4764221c91686dcf4d6b62d7a7b2d112e0b13/lua/dapui/windows/layout.lua#L47).

**Comment by rcarriga** [comment]
I do like the idea of the ability to extend the UI with custom windows that could be adapter or user specific :thinking: It's not really something that the internals are designed for, as you've seen. I'd like to refactor a bit before exposing the functionality.

**Comment by jedrzejboczar** [comment]
@rcarriga I noticed that version 3 of dap-ui added API for registering UI elements (`require('dapui').register_element`). However I wonder how one should use the API - when should I call `register_element`? Before or after `dapui.setup`? I tried both and none seems to work. When calling `register_element` before `dapui.setup` it will be [removed from elements at the start of dapui.setup](https://github.com/rcarriga/nvim-dap-ui/blob/master/lua/dapui/init.lua#L78-L84). On the other hand, calling `register_element` after `dapui.setup` won't call [windows.setup at the end of dapui.setup](https://github.com/rcarriga/nvim-dap-ui/blob/master/lua/dapui/init.lua#L179-L183). So I don't see a correct way to do this. For now I commented out the code at the beginning of `dapui.setup`.

I also have a question regaring the `dapui.Element`. I want to display an RTT terminal for [nvim-dap-cortex-debug](https://github.com/jedrzejboczar/nvim-dap-cortex-debug) but RTT is not connected yet when debugging have just started, so I just return a temporary scratch buffer. After a moment I receive a callback that RTT is connected, and from this point my `buffer` callback will return the actual RTT buffer. The problem is that dapui won't re-evaluate `buffer()` until I do some dap actions (stepping, etc.). Is there a way to force re-evaluation of `dapui.Element.buffer` manually (from my code)?

**Comment by rcarriga** [comment]
Ah I've changed the setup to only clear internal elements rather than all ones. I actually wanted to change the setup to allow custom elements be registered before or after the user configures dapui but for now only before will work.

> Is there a way to force re-evaluation of dapui.Element.buffer manually (from my code)?
```lua
local scratch_buf

local function set_win_buffer(actual_buf)
  local win = vim.fn.bufwinid(scratch_buf)
  if win == -1 then
    return
  end
  vim.api.nvim_win_set_buf(win, actual_buf)
end
```
That should wok right?

**Comment by jedrzejboczar** [comment]
Thanks, now `register_element` works when called before `dapui.setup`. 

> That should wok right?

This is a workaround that I was using, but I though there may be some dedicated API to do this. This solution works (sets correct buffer for dap-ui RTT window) but it replaces my current window (with source code) with the scratch buffer that was displayed in the RTT window before replacing it. 

It looks like this happens because dap-ui is guarding against changing the buffers displayed in dap-ui windows. E.g. I have dap-ui window A with buffer 1 and non-dap-ui window B with buffer 2. If I try to set A.buf = 3 (when B is current window), then as a result I will have A.buf=1, B.buf=3. How could I prevent this from happening?

**Comment by rcarriga** [comment]
That shouldn't be happening if the element's `buffer` function returns the new buffer before changing the window's buffer, can you confirm if that's happening?

**Comment by jedrzejboczar** [comment]
Yes, I checked that the order is:
* my `buffer` keeps returing temporary buffer 24
* an RTT channel is opened => new RTT terminal buffer 29 is created 
* my `buffer` keeps returning buffer 29
* `on_rtt_connect` callback is received, which detects that temporary buffer 24 is being displayed in a valid window 1021 => it finds the terminal buffer 29 and sets it for the window 1021

As a result I see buffer 29 in window 1021 as expected, but buffer 24 shows up in the source code window.

I don't really think that this might cause any problems in this situation, but just in case: the terminal buffer is created like this: 
1. `vim.api.nvim_create_buf`
2. `vim.api.nvim_open_win`: open temporary floating window
3. `vim.api.nvim_open_term`: attach terminal to the buffer (according to nvim docs this should be done in a buffer that is already displayed in a window)
4. `vim.api.nvim_win_close`: close temporary floating window
5. now `buffer` starts returning the terminal buffer
6. long after this `on_rtt_connect` is called

**Comment by jedrzejboczar** [comment]
Just checked that after setting `force_buffers = false` this does not happen. But I think it would be nice if custom elements could work correctly regardless of `force_buffers` option being set or not.

**Comment by jedrzejboczar** [comment]
One more thing (a bit unrelated): 

I noticed that the dap-ui buffers are created on nvim startup even if dap-ui has not yet been started. This was causing a problem when I tried loading new session using my session manager plugin, because it was configured  to delete all buffers before loading new session - so dap-ui had errors due to non-existing buffers. 

I didn't even have time to report this problem and today I tested with https://github.com/rcarriga/nvim-dap-ui/commit/110193102b4840be8bda6eb2358367026145ae3a and the problem is fixed, thanks! But if dap-ui is now implemented in such a way that the buffers are created on demand, wouldn't it be better to avoid creating these buffers on nvim startup? Not that it is any big issue, but it I think it might be a bit cleaner to delay buffer creation until they are needed.
